src/actions/ApplyAcl.ts

```diff
@@ -31,6 +35,23 @@ import { ERROR_KIND_FATAL, ERROR_KIND_PERMISSION } from "../ErrorCache";
  * @param {Mjolnir} mjolnir The Mjolnir client to apply the ACLs with.
  */
 export async function applyServerAcls(lists: BanList[], roomIds: string[], mjolnir: Mjolnir): Promise<RoomUpdateError[]> {
+    // There's probably a name for what this actually is, but we only want one instance of `_applyServerAcls` to be running at any time.
+    // So we naively promise chain them.
+    let timeoutId: NodeJS.Timeout|null = null;
+    inProgress = Promise.race([
+        new Promise((resolve, reject) => {
+            inProgress.finally(() => {
+                _applyServerAcls(lists, roomIds, mjolnir).then(resolve).catch(reject)
+            })
+        }),
+        // We don't want to stop ACL being applied if for some reason one gets stuck for 10 minutes.
+        new Promise((_, reject) => timeoutId = setTimeout(() => reject(new Error("Unable to apply ACL within a reasonable time.")), 600_000))
```

Comment:
Please make this a constant. Ideally, configurable but that can wait for a followup.